File: app/src/main/java/com/glucocontrol/data/remote/dto/GlucoseReadingDto.kt
package com.glucocontrol.data.remote.dto

import com.glucocontrol.domain.model.GlucoseReading
import com.glucocontrol.domain.model.ReadingTag
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.LocalDate
import java.time.LocalTime

/** DTO para lecturas leídas desde Supabase (incluye id generado por la BD). */
@Serializable
data class GlucoseReadingDto(
    val id: Long,
    @SerialName("user_id") val userId: String,
    @SerialName("value_mg_dl") val valueMgDl: Int,
    @SerialName("date_epoch_day") val dateEpochDay: Long,
    @SerialName("time_seconds_of_day") val timeSecondsOfDay: Int? = null,
    val tag: String = ReadingTag.OTRO.name,
    val notes: String? = null,
) {
    fun toDomain() = GlucoseReading(
        id = id,
        valueMgDl = valueMgDl,
        date = LocalDate.ofEpochDay(dateEpochDay),
        time = timeSecondsOfDay?.let { LocalTime.ofSecondOfDay(it.toLong()) },
        // Enum desconocido cae a OTRO para no romper lecturas antiguas
        tag = runCatching { ReadingTag.valueOf(tag) }.getOrDefault(ReadingTag.OTRO),
        notes = notes,
    )
}

/** DTO para INSERT: sin id (lo genera Supabase con BIGSERIAL). */
@Serializable
data class GlucoseReadingCreateDto(
    @SerialName("user_id") val userId: String,
    @SerialName("value_mg_dl") val valueMgDl: Int,
    @SerialName("date_epoch_day") val dateEpochDay: Long,
    @SerialName("time_seconds_of_day") val timeSecondsOfDay: Int? = null,
    val tag: String,
    val notes: String? = null,
) {
    companion object {
        fun fromDomain(reading: GlucoseReading, userId: String) = GlucoseReadingCreateDto(
            userId = userId,
            valueMgDl = reading.valueMgDl,
            dateEpochDay = reading.date.toEpochDay(),
            timeSecondsOfDay = reading.time?.toSecondOfDay(),
            tag = reading.tag.name,
            notes = reading.notes,
        )
    }
}

/** DTO para UPDATE: incluye id para que PostgREST identifique la fila. */
@Serializable
data class GlucoseReadingUpdateDto(
    val id: Long,
    @SerialName("user_id") val userId: String,
    @SerialName("value_mg_dl") val valueMgDl: Int,
    @SerialName("date_epoch_day") val dateEpochDay: Long,
    @SerialName("time_seconds_of_day") val timeSecondsOfDay: Int? = null,
    val tag: String,
    val notes: String? = null,
) {
    companion object {
        fun fromDomain(reading: GlucoseReading, userId: String) = GlucoseReadingUpdateDto(
            id = reading.id,
            userId = userId,
            valueMgDl = reading.valueMgDl,
            dateEpochDay = reading.date.toEpochDay(),
            timeSecondsOfDay = reading.time?.toSecondOfDay(),
            tag = reading.tag.name,
            notes = reading.notes,
        )
    }
}

File: lib/data/dto/glucose_reading_dto.dart
import 'package:glucocontrol/domain/model/glucose_reading.dart';
import 'package:glucocontrol/domain/model/reading_tag.dart';

// Columnas de la tabla Supabase — deben coincidir exactamente con el schema
class GlucoseReadingDto {
  final int id;
  final String userId;
  final int valueMgDl;
  final int dateEpochDay;
  final int? timeSecondsOfDay;
  final String tag;
  final String? notes;

  const GlucoseReadingDto({
    required this.id,
    required this.userId,
    required this.valueMgDl,
    required this.dateEpochDay,
    this.timeSecondsOfDay,
    required this.tag,
    this.notes,
  });

  factory GlucoseReadingDto.fromJson(Map<String, dynamic> json) =>
      GlucoseReadingDto(
        id: json['id'] as int,
        userId: json['user_id'] as String,
        valueMgDl: json['value_mg_dl'] as int,
        dateEpochDay: json['date_epoch_day'] as int,
        timeSecondsOfDay: json['time_seconds_of_day'] as int?,
        tag: (json['tag'] as String?) ?? 'OTRO',
        notes: json['notes'] as String?,
      );

  GlucoseReading toDomain() => GlucoseReading(
        id: id,
        date: DateTime.fromMillisecondsSinceEpoch(
          dateEpochDay * 86400 * 1000,
          isUtc: true,
        ),
        time: timeSecondsOfDay != null
            ? DateTime.fromMillisecondsSinceEpoch(0, isUtc: true).add(
                Duration(seconds: timeSecondsOfDay!),
              )
            : null,
        valueMgDl: valueMgDl,
        tag: ReadingTag.fromString(tag),
        notes: notes,
      );
}

class GlucoseReadingCreateDto {
  final String userId;
  final int valueMgDl;
  final int dateEpochDay;
  final int? timeSecondsOfDay;
  final String tag;
  final String? notes;

  const GlucoseReadingCreateDto({
    required this.userId,
    required this.valueMgDl,
    required this.dateEpochDay,
    this.timeSecondsOfDay,
    required this.tag,
    this.notes,
  });

  factory GlucoseReadingCreateDto.fromDomain(
    GlucoseReading r,
    String userId,
  ) {
    final epochDay = r.date
            .toUtc()
            .millisecondsSinceEpoch ~/
        86400000;
    final timeSeconds = r.time != null
        ? r.time!.hour * 3600 + r.time!.minute * 60 + r.time!.second
        : null;
    return GlucoseReadingCreateDto(
      userId: userId,
      valueMgDl: r.valueMgDl,
      dateEpochDay: epochDay,
      timeSecondsOfDay: timeSeconds,
      tag: r.tag.name.toUpperCase(),
      notes: r.notes,
    );
  }

  Map<String, dynamic> toJson() => {
        'user_id': userId,
        'value_mg_dl': valueMgDl,
        'date_epoch_day': dateEpochDay,
        if (timeSecondsOfDay != null) 'time_seconds_of_day': timeSecondsOfDay,
        'tag': tag,
        if (notes != null) 'notes': notes,
      };
}

class GlucoseReadingUpdateDto extends GlucoseReadingCreateDto {
  final int id;

  const GlucoseReadingUpdateDto({
    required this.id,
    required super.userId,
    required super.valueMgDl,
    required super.dateEpochDay,
    super.timeSecondsOfDay,
    required super.tag,
    super.notes,
  });

  factory GlucoseReadingUpdateDto.fromDomain(
    GlucoseReading r,
    String userId,
  ) {
    final base = GlucoseReadingCreateDto.fromDomain(r, userId);
    return GlucoseReadingUpdateDto(
      id: r.id,
      userId: base.userId,
      valueMgDl: base.valueMgDl,
      dateEpochDay: base.dateEpochDay,
      timeSecondsOfDay: base.timeSecondsOfDay,
      tag: base.tag,
      notes: base.notes,
    );
  }

  @override
  Map<String, dynamic> toJson() => {
        ...super.toJson(),
        'id': id,
      };
}
